apps/test/authoredHintUtilsTest.js

```diff
@@ -0,0 +1,255 @@
+        final_activity_id: 0,
+        final_attempt: 1,
+        final_level_source_id: 2
+      }, {
+        id: 'second',
+        prev_test_result: 'something okay',
+        next_test_result: 'something fine',
+        final_test_result: 'something rather poor',
+        final_time: 'something standard',
+        final_activity_id: 0,
+        final_attempt: 1,
+        final_level_source_id: 2
+      }]);
+    });
+
+  });
+
+  describe('recordUnfinishedHint', function () {
+    it('if no last_attempt_record is set, simply records the given value', function () {
+      authoredHintUtils.recordUnfinishedHint(hintOne);
+      var unfinishedHints = authoredHintUtils.getUnfinishedHints_();
+      assert.deepEqual(unfinishedHints, [hintOne]);
+    });
+
+    it('if last_attempt_record is set, extends the given hint without overriding', function () {
+      localStorage.setItem('last_attempt_record', JSON.stringify(record));
+
+      authoredHintUtils.recordUnfinishedHint(hintOne);
+      var unfinishedHints = authoredHintUtils.getUnfinishedHints_();
+
+      assert.deepEqual(unfinishedHints, [{
+        id: 'first',
+        prev_time: 'something decent',
+        next_time: 'something great',
+        final_time: 'something awesome',
+        prev_test_result: 'something grand',
+        prev_activity_id: 0,
+        prev_attempt: 1,
+        prev_level_source_id: 2
+      }]);
+    });
+
+
+  });
+
+  describe('finishHints', function () {
+
+    it('sets last_attempt_record', function () {
+      assert.isNull(localStorage.getItem('last_attempt_record'));
+      authoredHintUtils.finishHints(record);
+      assert.equal(localStorage.getItem('last_attempt_record'), JSON.stringify(record));
+    });
+
+    it('clears unfinished_hints', function () {
+      authoredHintUtils.recordUnfinishedHint(hintOne);
+      assert.equal(localStorage.getItem('unfinished_authored_hint_views'), JSON.stringify([hintOne]));
+      authoredHintUtils.finishHints(record);
+      assert.equal(localStorage.getItem('unfinished_authored_hint_views'), JSON.stringify([]));
+    });
+
+    it('extends unfinished_hints and moves them to finished_hints', function () {
+      authoredHintUtils.recordUnfinishedHint(hintOne);
+      assert.equal(localStorage.getItem('unfinished_authored_hint_views'), JSON.stringify([hintOne]));
+      authoredHintUtils.finishHints(record);
+      var finishedHints = authoredHintUtils.getFinishedHints_();
+
+      assert.deepEqual(finishedHints, [{
+        id: 'first',
+        prev_time: 'something decent',
+        next_time: 'something great',
+        final_time: 'something awesome',
+        next_test_result: 'something grand',
+        next_activity_id: 0,
+        next_attempt: 1,
+        next_level_source_id: 2
+      }]);
+    });
+
+  });
+
+  describe('submitHints', function () {
+
+    it('finishes all unfinished hints using the last hint', function () {
+      $.ajax = function () {};
```

Comment:
for tests where we're modifying $.ajax, should we have some before/after step that stashes/restores the original state of $.ajax?